src/goto-instrument/contracts/dynamic-frames/dfcc_instrument.cpp

```diff
@@ -0,0 +1,1359 @@
+        check_var,
+        library.dfcc_fun_symbol[dfcc_funt::WRITE_SET_CHECK_ARRAY_SET]
+          .symbol_expr(),
+        {write_set, dest}},
+      target_location));
+
+    // add property class on assertion source_location
+    source_locationt check_location(target_location);
+    check_location.set_property_class("assigns");
+    std::string comment = "Check that array_set(" +
+                          from_expr_using_mode(ns, mode, dest) +
+                          ", ...) is allowed by the assigns clause";
+    check_location.set_comment(comment);
+
+    payload.add(goto_programt::make_assertion(check_var, check_location));
+    payload.add(goto_programt::make_dead(check_var));
+
+    auto label_instruction =
+      payload.add(goto_programt::make_skip(target_location));
+    goto_instruction->complete_goto(label_instruction);
+
+    insert_before_swap_and_advance(goto_program, target, payload);
+  }
+  else if(statement == ID_array_copy)
+  {
+    // ```
+    // IF !write_set GOTO skip_target;
+    // DECL check_array_copy: bool;
+    // CALL check_array_copy = check_array_copy(write_set, dest);
+    // ASSERT(check_array_copy);
+    // DEAD check_array_copy;
+    // skip_target: SKIP;
+    // ----
+    // OTHER {statemet = array_copy, args = {dest, src}};
+    // ```
+    const auto &mode = utils.get_function_symbol(function_id).mode;
+    goto_programt payload;
+
+    auto goto_instruction = payload.add(goto_programt::make_incomplete_goto(
+      utils.make_null_check_expr(write_set)));
+
+    auto &check_sym = get_fresh_aux_symbol(
+      bool_typet(),
+      id2string(function_id),
+      "__check_array_copy",
+      target_location,
+      mode,
+      goto_model.symbol_table);
+
+    const auto &check_var = check_sym.symbol_expr();
+
+    payload.add(goto_programt::make_decl(check_var));
+
+    const auto &dest = target->get_other().operands().at(0);
+
+    payload.add(goto_programt::make_function_call(
+      code_function_callt{
+        check_var,
+        library.dfcc_fun_symbol[dfcc_funt::WRITE_SET_CHECK_ARRAY_COPY]
+          .symbol_expr(),
+        {write_set, dest}},
+      target_location));
+
+    // add property class on assertion source_location
+    source_locationt check_location(target_location);
+    check_location.set_property_class("assigns");
+    std::string comment = "Check that array_copy(" +
+                          from_expr_using_mode(ns, mode, dest) +
+                          ", ...) is allowed by the assigns clause";
+    check_location.set_comment(comment);
+
+    payload.add(goto_programt::make_assertion(check_var, check_location));
+    payload.add(goto_programt::make_dead(check_var, target_location));
+
+    auto label_instruction =
+      payload.add(goto_programt::make_skip(target_location));
+    goto_instruction->complete_goto(label_instruction);
+
+    insert_before_swap_and_advance(goto_program, target, payload);
+  }
+  else if(statement == ID_array_replace)
```

Comment:
I cannot tell for sure, but isn't this branch almost the same as `ID_array_copy` (except for the symbol)?

Reply:
factored